vespalib/src/tests/hwaccelrated/hwaccelrated_test.cpp

```diff
@@ -15,26 +17,33 @@ std::vector<T> createAndFill(size_t sz) {
     return v;
 }
 
-template<typename T>
-void verifyEuclideanDistance(const hwaccelrated::IAccelrated & accel) {
-    const size_t testLength(255);
+template<typename T, typename P>
+void verifyEuclideanDistance(const hwaccelrated::IAccelrated & accel, size_t testLength) {
     srand(1);
     std::vector<T> a = createAndFill<T>(testLength);
     std::vector<T> b = createAndFill<T>(testLength);
     for (size_t j(0); j < 0x20; j++) {
-        T sum(0);
+        P sum(0);
         for (size_t i(j); i < testLength; i++) {
-            sum += (a[i] - b[i]) * (a[i] - b[i]);
+            P d = P(a[i]) - P(b[i]);
+            sum += d * d;
         }
-        T hwComputedSum(accel.squaredEuclideanDistance(&a[j], &b[j], testLength - j));
+        P hwComputedSum(accel.squaredEuclideanDistance(&a[j], &b[j], testLength - j));
         EXPECT_EQUAL(sum, hwComputedSum);
     }
 }
 
+void
+verifyEuclideanDistance(const hwaccelrated::IAccelrated & accelrator, size_t testLength) {
+    verifyEuclideanDistance<int8_t, float>(accelrator, testLength);
+    verifyEuclideanDistance<float, float>(accelrator, testLength);
+    verifyEuclideanDistance<double, float>(accelrator, testLength);
+}
+
 TEST("test euclidean distance") {
     hwaccelrated::GenericAccelrator genericAccelrator;
-    verifyEuclideanDistance<float>(genericAccelrator);
-    verifyEuclideanDistance<double >(genericAccelrator);
+    verifyEuclideanDistance(hwaccelrated::GenericAccelrator(),255);
+    verifyEuclideanDistance(hwaccelrated::IAccelrated::getAccelerator(),255);
```

Comment:
We should test all supported accelerators, not just the generic and 'best' ones. Ideally, we should also make sure that the build CPUs support 'all' accelerators, to get maximum test coverage.

Reply:
When an application starts it tests the accelrator chosen for the cpu against the generic one.
It is suboptimal, but the best I can think of. We will never be able to test all anyway.